### ISSUE TITLE
Sourcemaps: Comment lines are not mapped

### ISSUE THREAD

**Issue by turbo** [jashkenas/coffeescript#4772]
It seems that comment lines are not mapped within the sourcemap. Here's how I tested it:

<details><summary>File: <i>source.coffee</i></summary><p>

```coffeescript
fs = require 'fs'

###

Important comment

###

console.log 'hello'
```
</p></details>  
<br>

Compiled with `coffee --bare --no-header -cm source.coffee` yields

<details><summary>File: <i>source.js</i></summary><p>

```js
var fs;

fs = require('fs');

/*

Important comment

*/
console.log('hello');

//# sourceMappingURL=source.js.map
```
</p></details>  
<br>

Now trying to resolve the console log line will be successful, but resolving the comment line just yields null:

```coffeescript
sm = require 'source-map'
fs = require 'fs'

smc = new sm.SourceMapConsumer fs.readFileSync "./mapstest/source.js.map", 'utf8'

console.log smc.originalPositionFor
  line: 10,
  column: 1

# Gives correct answer: { source: 'source.coffee', line: 9, column: 0, name: null }

console.log smc.originalPositionFor
  line: 7,
  column: 1

# Is not resolved: { source: null, line: null, column: null, name: null }
```

CS does a great job of inserting the comments where they belong, which is important for e.g. flow typing, so I think the should be mapped, too. If that's even possible that is.

**Comment by GeoffreyBooth** [comment]
I did a little experiment. I took your `source.js` and saved a copy without the last-line comment that referenced a source map. I ran that through Babel, telling Babel to generate a source map:

```bash
npx babel-cli source.js --out-file babel.js --source-maps
```

The resulting `babel.js` is identical to your `source.js`, aside from Babel adding `//# sourceMappingURL=babel.js.map` to the bottom of `babel.js`. All well and good. The generated `babel.js.map` looks like this (formatted):

```json
{
  "version": 3,
  "sources": [
    "source.js"
  ],
  "names": [],
  "mappings": "AAAA,IAAI,EAAJ;;AAEA,KAAK,QAAQ,IAAR,CAAL;;AAEA;;;;;AAKA,QAAQ,GAAR,CAAY,OAAZ",
  "file": "babel.js",
  "sourcesContent": [
    "var fs;\n\nfs = require('fs');\n\n/*\n\nImportant comment\n\n*/\nconsole.log('hello');\n"
  ]
}
```

Then I ran your test code against this source map. The line numbers are the same for `Important comment` and `console.log('hello');`. I’ve updated with the responses for running it against `babel.js.map`:

```coffee
sm = require 'source-map'
fs = require 'fs'

smc = new sm.SourceMapConsumer fs.readFileSync './babel.js.map', 'utf8'

console.log smc.originalPositionFor
  line: 10,
  column: 1

# Gives correct answer: { source: 'source.js', line: 10, column: 0, name: null }

console.log smc.originalPositionFor
  line: 7,
  column: 1

# Is not resolved: { source: null, line: null, column: null, name: null }
```

So in other words, it appears that Babel doesn’t create source map mappings for comments either.

This feels correct to me. The purpose of source maps is to provide places for a debugger to set a breakpoint. You can’t set a breakpoint on a comment, or on whitespace (the other thing that source maps ignore), so there’s no point in creating mappings for such nodes or lines. Also, there’s a premium placed on keeping source maps small, hence the `AAAA,IAAI` etc. condensed encoding. Thus it follows that source maps would skip unnecessary mappings.

So basically, I don’t see the purpose of CoffeeScript source maps mapping comments. Source maps aren’t a reflection of what’s “important” in a file.

It would also be a huge challenge for the CoffeeScript compiler to map comments, as comments aren’t nodes like normal code; they’re _properties_ of nearby nodes, so they don’t go through the same parsing process as real nodes do and therefore don’t have location data (start and end column/row numbers). They’re output via the same process where extra whitespace is generated, for example to add blank lines between functions and that kind of thing.

**Comment by turbo** [comment]
First, thanks for the thorough answer.

> This feels correct to me. The purpose of source maps is to provide places for a debugger to set a breakpoint. You can’t set a breakpoint on a comment, or on whitespace (the other thing that source maps ignore), so there’s no point in creating mappings for such nodes or lines. 

There might be. I've created a custom frontend to flow's static analysis which will show errors and annoations using the original coffee source file. Example:

```text
✖ Type check failed, 2 errors found.

✖ (0)
|  ℹ Reference: object literal in 'src/test.coffee':
|
|    12: fn 'hello', {num: "3"}
|                    🡱
|
|  Error: This type is incompatible with the expected param type of
|
|  ℹ Reference: object type in 'src/test.coffee':
|
|     9: fn = (str ###: string ###, obj ###: Obj ###) ###: string ### ->
|                                   🡱
|    10:   str + obj.num
|
```

So far, so good. However, flow outputs additional context using the `extra` messages. Since custom type defs are only possible in comments, e.g.

```coffeescript
###::
type Obj = {
  num: number,
};
###
```

the context will point to this definition (flow doesn't care if that's a comment or not) in the result file:

```js
/*::
type Obj = {
  num: number,
};
*/
```

However, my frontend can't resolve the definition since comments are not mapped. Flow isn't the only example for this, significant comments are also a problem when doc tools find an error (think JSDoc or apidoc).

> It would also be a huge challenge for the CoffeeScript compiler to map comments, as comments aren’t nodes like normal code; they’re properties of nearby nodes, so they don’t go through the same parsing process as real nodes do and therefore don’t have location data (start and end column/row numbers). They’re output via the same process where extra whitespace is generated, for example to add blank lines between functions and that kind of thing.

Interesting. So I guess a way to do this then would be to introduce syntax into the compiler which will output mapped type annotations. Is the wiki on how to hack on the compiler still up to date? I really want to try this.

**Comment by GeoffreyBooth** [comment]
I think rather than hacking the compiler, a better approach would be for your tool to add a postprocessing step where it searches both the Coffee and JS files for comments and creates a new source map of just those comments; and then merges this new source map with the CoffeeScript-generated one. It shouldn't be hard to get row/column coordinates for `###` and `/*`/`*/` strings, and `source-map` can convert your data into a source map as well as merge two maps together.

**Comment by turbo** [comment]
That's a good idea for the comment based workflow, however inline comments for type annotations is clunky IMO. My end goal is to have proper gradual, static typing support - for that I will need to look into the compiler anyway. There's more syntax to flow than "just" typedefs. Importing types has to be supported, too.

I guess the compiler could still emit ES with commented type annotations in order for the output to be executable unmodified, but be valid flow input, too.

**Comment by GeoffreyBooth** [comment]
I would encourage you to avoid hacking the compiler if at all possible. Why not have your tool take a pass at the code both before and after the CoffeeScript compiler does? Something like:

1. Convert the input code into valid CoffeeScript using `###` comments.
1. Pass that through the CoffeeScript compiler.
1. Do the stuff described above, adding comments to the source map.

Then you just need to decide what your input would be, that can be reliably converted into CoffeeScript comments. Maybe `::` followed by a space, then a word? So `:: String`, wherever it appears, becomes `###:: String ###`.

**Comment by turbo** [comment]
In order to handle all edge cases provided by a new syntax element (again, it's not just type annotations, but also definitions), I'd have to write a parser which understands what I want, and what to exclude because it's coffeescript.

That will be a mess. [I went down this route before](https://github.com/turbo/OOPEAu3), extending a language by means of preprocessing. Writing a parser, even just a simple rec. descent like I did for my other [compiler](https://github.com/turbo/Perseus) would have been easier.

So why create a new parser and glue something to the front and back, instead of extending the language itself? Sure, a quick and dirty solution will probably cover 80% of the use cases, but I want to create something robust and usable. It seems to have been a valid strategy for e.g. contracts.coffee.

I'm thankful for every help and reference I can get on how to work on the compiler source.
